File: Src/Syntax/Syntax/Stmts.cs
using Citron.Syntax;
using System;
using System.Collections.Generic;
using Citron.Collections;
using System.Diagnostics;
using System.Linq;
using System.Xml.XPath;
using System.Data;

namespace Citron.Syntax
{
    public abstract record class Stmt : ISyntaxNode
    {
        public abstract TResult Accept<TVisitor, TResult>(ref TVisitor visitor)
            where TVisitor : struct, IStmtVisitor<TResult>;
    }

    // if, for 등에 나오는 한줄짜리 Stmt
    public abstract record class EmbeddableStmt
    {
        public record class Single(Stmt Stmt) : EmbeddableStmt;
        public record class Multiple(ImmutableArray<Stmt> Stmts) : EmbeddableStmt;
    }

    // 명령어
    // TODO: commands의 Length가 1인 contract를 추가하자
    public record CommandStmt(ImmutableArray<StringExp> Commands) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitCommand(this);
    }

    // int a = 0, b, c;
    public record class VarDeclStmt(VarDecl VarDecl) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitVarDecl(this);
    }

    // if ($cond) $body else $ElseBody
    public record class IfStmt(Exp Cond, EmbeddableStmt Body, EmbeddableStmt? ElseBody) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIf(this);
    }

    // if (testType varName = exp
    public record class IfTestStmt(TypeExp TestTypeExp, string VarName, Exp Exp, EmbeddableStmt Body, EmbeddableStmt? ElseBody) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitIfTest(this);
    }

    public record class ForStmt(ForStmtInitializer? Initializer, Exp? CondExp, Exp? ContinueExp, EmbeddableStmt Body) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitFor(this);
    }

    public record ContinueStmt : Stmt
    {   
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitContinue(this);
    }

    public record class BreakStmt : Stmt
    {   
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBreak(this);
    }
    
    public record struct ReturnValueInfo(Exp Value);

    public record class ReturnStmt(ReturnValueInfo? Info) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitReturn(this);
    }

    // Stmt중간에 명시적으로 { }를 사용한 경우에만 BlockStmt로 나타낸다. 함수, for, async, await, if등에 나타나는 { 이후 구문들은 EmbeddableStmt를 사용한다
    public record class BlockStmt(ImmutableArray<Stmt> Stmts) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBlock(this);
    }
    
    public record class BlankStmt : Stmt
    {        
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitBlank(this);
    }

    public record class ExpStmt(Exp Exp) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitExp(this);
    }

    public record class TaskStmt(ImmutableArray<Stmt> Body) : Stmt // task { ... }
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitTask(this);
    }

    public record class AwaitStmt(ImmutableArray<Stmt> Body) : Stmt // await { ... }
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitAwait(this);
    }

    public record class AsyncStmt(ImmutableArray<Stmt> Body) : Stmt // async { ... }
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitAsync(this);
    }

    public record class ForeachStmt(TypeExp Type, string VarName, Exp Enumerable, EmbeddableStmt Body) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitForeach(this);
    }

    public record class YieldStmt(Exp Value) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitYield(this);
    }

    // Stmt에 사용되는 Directive랑 Decl-Level에서 사용되는 Directive가 다르므로 구분해도 될 것 같다
    public record class DirectiveStmt(string Name, ImmutableArray<Exp> Args) : Stmt
    {
        public override TResult Accept<TVisitor, TResult>(ref TVisitor visitor) => visitor.VisitDirective(this);
    }
}

File: src/Syntax/Stmts.cpp
#include "pch.h"
#include "Syntax/Stmts.h"

using namespace std;

namespace Citron::Syntax {

CommandStmt::CommandStmt(vector<StringExp> commands)
    : commands(std::move(commands))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(CommandStmt)

struct IfStmt::Impl
{
    Exp cond;
    EmbeddableStmt body; 
    optional<EmbeddableStmt> elseBody;
};

IfStmt::IfStmt(Exp cond, EmbeddableStmt body, optional<EmbeddableStmt> elseBody)
    : impl(new Impl { std::move(cond), std::move(body), std::move(elseBody) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(IfStmt)

Exp& IfStmt::GetCond()
{
    return impl->cond;
}

EmbeddableStmt& IfStmt::GetBody()
{
    return impl->body;
}

optional<EmbeddableStmt>& IfStmt::GetElseBody()
{
    return impl->elseBody;
}

struct IfTestStmt::Impl 
{
    TypeExp testTypeExp;
    string varName;
    Exp exp; 
    EmbeddableStmt body;
    optional<EmbeddableStmt> elseBody;
};

IfTestStmt::IfTestStmt(TypeExp testTypeExp, string varName, Exp exp, EmbeddableStmt body, optional<EmbeddableStmt> elseBody)
    : impl(new Impl { std::move(testTypeExp), std::move(varName), std::move(exp), std::move(body), std::move(elseBody) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(IfTestStmt)

TypeExp& IfTestStmt::GetTestTypeExp()
{
    return impl->testTypeExp;
}

string& IfTestStmt::GetVarName()
{
    return impl->varName;
}

Exp& IfTestStmt::GetExp()
{
    return impl->exp;
}

EmbeddableStmt& IfTestStmt::GetBody()
{
    return impl->body;
}

optional<EmbeddableStmt>& IfTestStmt::GetElseBody()
{
    return impl->elseBody;
}

struct ForStmt::Impl
{
    optional<ForStmtInitializer> initializer;
    optional<Exp> condExp;
    optional<Exp> continueExp;
    EmbeddableStmt body;
};

ForStmt::ForStmt(optional<ForStmtInitializer> initializer, optional<Exp> condExp, optional<Exp> continueExp, EmbeddableStmt body)
    : impl(new Impl { std::move(initializer), std::move(condExp), std::move(continueExp), std::move(body) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(ForStmt)

optional<ForStmtInitializer>& ForStmt::GetInitializer()
{
    return impl->initializer;
}

optional<Citron::Syntax::Exp>& ForStmt::GetCondExp()
{
    return impl->condExp;
}

optional<Citron::Syntax::Exp>& ForStmt::GetContinueExp()
{
    return impl->continueExp;
}

EmbeddableStmt& ForStmt::GetBody()
{
    return impl->body;
}

ReturnStmt::ReturnStmt(optional<ReturnValueInfo> info)
    : info(std::move(info))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(ReturnStmt)

BlockStmt::BlockStmt(vector<Stmt> stmts)
    : stmts(std::move(stmts))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(BlockStmt)

TaskStmt::TaskStmt(vector<Stmt> body)
    : body(std::move(body))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(TaskStmt)


AwaitStmt::AwaitStmt(vector<Stmt> body)
    : body(std::move(body))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(AwaitStmt)

AsyncStmt::AsyncStmt(vector<Stmt> body)
    : body(std::move(body))
{
}

IMPLEMENT_DEFAULTS_DEFAULT(AsyncStmt)

struct ForeachStmt::Impl
{
    TypeExp type;
    string varName;
    Exp enumerable;
    EmbeddableStmt body;
};

ForeachStmt::ForeachStmt(TypeExp type, string varName, Exp enumerable, EmbeddableStmt body)
    : impl(new Impl { std::move(type), std::move(varName), std::move(enumerable), std::move(body) })
{
}

IMPLEMENT_DEFAULTS_PIMPL(ForeachStmt)

TypeExp& ForeachStmt::GetType()
{
    return impl->type;
}

string& ForeachStmt::GetVarName()
{
    return impl->varName;
}

Exp& ForeachStmt::GetEnumerable()
{
    return impl->enumerable;
}

EmbeddableStmt& ForeachStmt::GetBody()
{
    return impl->body;
}

}